PartRenderingEngineTests.ensureCleanUpAddonCleansUp still fails randomly

https://github.com/eclipse-platform/eclipse.platform.ui/runs/57884315250

```
CleanupAddon should ensure that partStack is not rendered anymore, as all childs have been removed
java.lang.AssertionError: CleanupAddon should ensure that partStack is not rendered anymore, as all childs have been removed
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.eclipse.e4.ui.tests.workbench.PartRenderingEngineTests.ensureCleanUpAddonCleansUp(PartRenderingEngineTests.java:2472)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.eclipse.e4.ui.tests.rules.WorkbenchContextRule$1.evaluate(WorkbenchContextRule.java:62)
	Suppressed: java.lang.Throwable: Screenshot written to C:\Users\RUNNER~1\AppData\Local\Temp\org.eclipse.e4.ui.tests.workbench.PartRenderingEngineTests.ensureCleanUpAddonCleansUp.png
		at org.eclipse.test.Screenshots$ScreenshotOnFailure.failed(Screenshots.java:41)
AWT screenshot saved to: C:\Users\RUNNER~1\AppData\Local\Temp\org.eclipse.e4.ui.tests.workbench.PartRenderingEngineTests.ensureCleanUpAddonCleansUp.png
```

The test failure can be reproduced locally (on Linux at least) after executing the test ~30-70 times.

See:
- https://github.com/eclipse-platform/eclipse.platform.ui/commit/64500cdcaf3233b8f368674d01b39b010db7af67
- https://github.com/eclipse-platform/eclipse.platform.ui/commit/7a24db1a216ef1e9940840c7c1156e1f9a82fc93
- https://github.com/eclipse-platform/eclipse.platform.ui/commit/a3fa06583905e66550c0b4e8a3961320cfb4ca43
- https://github.com/eclipse-platform/eclipse.platform.ui/issues/751

--- a/bundles/org.eclipse.e4.ui.workbench.addons.swt/META-INF/MANIFEST.MF
+++ b/bundles/org.eclipse.e4.ui.workbench.addons.swt/META-INF/MANIFEST.MF
@@ -1,7 +1,7 @@
 Manifest-Version: 1.0
 Bundle-ManifestVersion: 2
 Bundle-SymbolicName: org.eclipse.e4.ui.workbench.addons.swt;singleton:=true
-Bundle-Version: 1.6.0.qualifier
+Bundle-Version: 1.6.100.qualifier
 Bundle-Name: %pluginName
 Bundle-Vendor: %providerName
 Bundle-Localization: plugin

--- a/bundles/org.eclipse.e4.ui.workbench.addons.swt/src/org/eclipse/e4/ui/workbench/addons/cleanupaddon/CleanupAddon.java
+++ b/bundles/org.eclipse.e4.ui.workbench.addons.swt/src/org/eclipse/e4/ui/workbench/addons/cleanupaddon/CleanupAddon.java
@@ -346,16 +346,13 @@ public class CleanupAddon {
 				int visCount = modelService.countRenderableChildren(container);
 
 				// Remove stacks with no visible children from the display (but not the
-				// model)
-				final MElementContainer<MUIElement> theContainer = container;
-				if (visCount == 0) {
-					Display.getCurrent().asyncExec(() -> {
-						int visCount1 = modelService.countRenderableChildren(theContainer);
-						if (!isLastEditorStack(theContainer) && visCount1 == 0) {
-							theContainer.setToBeRendered(false);
-							transferPrimaryDataStackIfRemoved(theContainer);
-						}
-					});
+				// model). Since event delivery is synchronous (EventBroker.send via
+				// EventAdmin.sendEvent and Display.syncExec), the visCount reflects
+				// the actual state at the time of this event and can be acted upon
+				// immediately without deferring via asyncExec.
+				if (visCount == 0 && !isLastEditorStack(container)) {
+					container.setToBeRendered(false);
+					transferPrimaryDataStackIfRemoved(container);
 				} else if (container.getParent() != null) { // omit detached windows
 					// if there are rendered elements but none are 'visible' we should
 					// make the container invisible as well

--- a/tests/org.eclipse.e4.ui.tests/META-INF/MANIFEST.MF
+++ b/tests/org.eclipse.e4.ui.tests/META-INF/MANIFEST.MF
@@ -2,7 +2,7 @@ Manifest-Version: 1.0
 Bundle-ManifestVersion: 2
 Bundle-Name: %pluginName
 Bundle-SymbolicName: org.eclipse.e4.ui.tests;singleton:=true
-Bundle-Version: 0.16.0.qualifier
+Bundle-Version: 0.16.100.qualifier
 Bundle-Vendor: %providerName
 Bundle-Localization: plugin
 Require-Bundle: org.eclipse.emf.ecore.xmi;bundle-version="2.4.0",

--- a/tests/org.eclipse.e4.ui.tests/src/org/eclipse/e4/ui/tests/workbench/PartRenderingEngineTests.java
+++ b/tests/org.eclipse.e4.ui.tests/src/org/eclipse/e4/ui/tests/workbench/PartRenderingEngineTests.java
@@ -64,7 +64,6 @@ import org.eclipse.swt.widgets.Display;
 import org.eclipse.swt.widgets.Shell;
 import org.eclipse.swt.widgets.ToolBar;
 import org.eclipse.swt.widgets.Widget;
-import org.eclipse.ui.tests.harness.util.DisplayHelper;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Disabled;
@@ -2447,12 +2446,9 @@ public class PartRenderingEngineTests {
 		assertTrue(partStackForPartBPartC.isToBeRendered(), " PartStack with children should be rendered");
 		partService.hidePart(partB);
 		partService.hidePart(partC);
-		// DisplayHelper.waitForCondition() handles event processing via Display.sleep()
-		// and retries. Calling spinEventLoop() here creates a race condition where
-		// events may be processed before CleanupAddon's asyncExec() is queued (line 352).
-		assertTrue(
-				DisplayHelper.waitForCondition(Display.getDefault(), 30_000,
-						() -> !partStackForPartBPartC.isToBeRendered()), "CleanupAddon should ensure that partStack is not rendered anymore, as all childs have been removed");
+		contextRule.spinEventLoop();
+		assertFalse(partStackForPartBPartC.isToBeRendered(),
+				"CleanupAddon should ensure that partStack is not rendered anymore, as all childs have been removed");
 		// PartStack with IPresentationEngine.NO_AUTO_COLLAPSE should not be removed
 		// even if children are removed
 		partService.hidePart(editor, true);

